From f1b29cf1a4e49184c9d3709afebd4f9996ea949a Mon Sep 17 00:00:00 2001
From: Vladimir Oltean <olteanv@gmail.com>
Date: Tue, 2 Oct 2018 00:54:17 +0300
Subject: [PATCH 1/2] Allow webview_zygote to read /dev/ion

* I don't have a good intuition of whether this change is good or bad,
  however regular Zygote has this permission, and Webview processes fail
  to load in enforcing mode without this change.
* TEST: Open Gmail on a msm8976 tablet.

I WebViewFactory: Loading com.android.webview version 66.0.3359.158 (code 336015855)
W cr_ChildProcLH: Create a new ChildConnectionAllocator with package name = com.android.webview, sandboxed = true
W libprocessgroup: kill(-4305, 9) failed: No such process
I libprocessgroup: Successfully killed process cgroup uid 10064 pid 4305 in 95ms
I cr_BrowserStartup: Initializing chromium process, singleProcess=false
I cr_base : Android Locale: en_US requires .pak files: []
E webview_zygote: Unable to restat fd 28: Permission denied
F webview_zygote: jni_internal.cc:616] JNI FatalError called: (com.android.webview:sandboxed_process0) Unable to stat 28
W webview_zygote: type=1400 audit(0.0:27): avc: denied { getattr } for path="/dev/ion" dev="tmpfs" ino=12718 scontext=u:r:webview_zygote:s0 tcontext=u:object_r:ion_device:s0 tclass=chr_file permissive=0
W webview_zygote: type=1400 audit(0.0:28): avc: denied { getattr } for path="/dev/ion" dev="tmpfs" ino=12718 scontext=u:r:webview_zygote:s0 tcontext=u:object_r:ion_device:s0 tclass=chr_file permissive=0
W webview_zygote: type=1400 audit(0.0:29): avc: denied { read } for name="app_process32" dev="mmcblk0p25" ino=467 scontext=u:r:webview_zygote:s0 tcontext=u:object_r:zygote_exec:s0 tclass=file permissive=0
F webview_zygote: runtime.cc:558] Runtime aborting...
F webview_zygote: runtime.cc:558] Dumping all threads without appropriate locks held: thread list lock mutator lock
F webview_zygote: runtime.cc:558] All threads:
F webview_zygote: runtime.cc:558] DALVIK THREADS (1):
F webview_zygote: runtime.cc:558] "main" prio=5 tid=1 Runnable
F webview_zygote: runtime.cc:558]   | group="" sCount=0 dsCount=0 flags=0 obj=0x74d7cf48 self=0xf2f73000
F webview_zygote: runtime.cc:558]   | sysTid=2332 nice=0 cgrp=default sched=0/0 handle=0xf6b54494
F webview_zygote: runtime.cc:558]   | state=R schedstat=( 80557599 15727867 94 ) utm=5 stm=3 core=5 HZ=100
F webview_zygote: runtime.cc:558]   | stack=0xff72b000-0xff72d000 stackSize=8MB
F webview_zygote: runtime.cc:558]   | held mutexes= "abort lock" "mutator lock"(shared held)
F webview_zygote: runtime.cc:558]   native: #00 pc 002d975f  /system/lib/libart.so (art::DumpNativeStack(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, int, BacktraceMap*, char const*, art::ArtMethod*, void*, bool)+134)
F webview_zygote: runtime.cc:558]   native: #01 pc 0036e98b  /system/lib/libart.so (art::Thread::DumpStack(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, bool, BacktraceMap*, bool) const+210)
F webview_zygote: runtime.cc:558]   native: #02 pc 0036b143  /system/lib/libart.so (art::Thread::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, bool, BacktraceMap*, bool) const+34)
F webview_zygote: runtime.cc:558]   native: #03 pc 00383df9  /system/lib/libart.so (art::DumpCheckpoint::Run(art::Thread*)+624)
F webview_zygote: runtime.cc:558]   native: #04 pc 0037e0df  /system/lib/libart.so (art::ThreadList::RunCheckpoint(art::Closure*, art::Closure*)+314)
F webview_zygote: runtime.cc:558]   native: #05 pc 0037d7d7  /system/lib/libart.so (art::ThreadList::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, bool)+758)
F webview_zygote: runtime.cc:558]   native: #06 pc 0034d8fb  /system/lib/libart.so (art::Runtime::Abort(char const*)+314)
F webview_zygote: runtime.cc:558]   native: #07 pc 000071b3  /system/lib/libbase.so (android::base::LogMessage::~LogMessage()+494)
F webview_zygote: runtime.cc:558]   native: #08 pc 00265a0f  /system/lib/libart.so (art::JNI::FatalError(_JNIEnv*, char const*)+122)
F webview_zygote: runtime.cc:558]   native: #09 pc 0010b4d5  /system/lib/libandroid_runtime.so (_ZZN12_GLOBAL__N_123ForkAndSpecializeCommonEP7_JNIEnvjjP10_jintArrayiP13_jobjectArrayxxiP8_jstringS7_bS3_S3_bS7_S7_ENK3$_0clERKNSt3__112basic_stringIcNS9_11char_traitsIcEENS9_9allocatorIcEEEE+92)
F webview_zygote: runtime.cc:558]   native: #10 pc 0010b28f  /system/lib/libandroid_runtime.so ((anonymous namespace)::ForkAndSpecializeCommon(_JNIEnv*, unsigned int, unsigned int, _jintArray*, int, _jobjectArray*, long long, long long, int, _jstring*, _jstring*, bool, _jintArray*, _jintArray*, bool, _jstring*, _jstring*)+4958)
F webview_zygote: runtime.cc:558]   native: #11 pc 001097f1  /system/lib/libandroid_runtime.so (android::com_android_internal_os_Zygote_nativeForkAndSpecialize(_JNIEnv*, _jclass*, int, int, _jintArray*, int, _jobjectArray*, int, _jstring*, _jstring*, _jintArray*, _jintArray*, unsigned char, _jstring*, _jstring*)+476)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.Zygote.nativeForkAndSpecialize(Native method)
W VideoCapabilities: Unrecognized profile/level 0/3 for video/mpeg2
F webview_zygote: runtime.cc:558]   at com.android.internal.os.Zygote.forkAndSpecialize(Zygote.java:139)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.ZygoteConnection.processOneCommand(ZygoteConnection.java:234)
E chromium: [ERROR:devtools_http_handler.cc(292)] Cannot start http server for devtools. Stop devtools.
F webview_zygote: runtime.cc:558]   at com.android.internal.os.ZygoteServer.runSelectLoop(ZygoteServer.java:204)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.WebViewZygoteInit.main(WebViewZygoteInit.java:160)
F webview_zygote: runtime.cc:558]   at java.lang.reflect.Method.invoke(Native method)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
D AutofillManagerServiceImpl: Reset component for user 0 ()
F webview_zygote: runtime.cc:558]   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
F webview_zygote: runtime.cc:558]
F webview_zygote: runtime.cc:558] Aborting thread:
F webview_zygote: runtime.cc:558] "main" prio=5 tid=1 Native
F webview_zygote: runtime.cc:558]   | group="" sCount=0 dsCount=0 flags=0 obj=0x74d7cf48 self=0xf2f73000
F webview_zygote: runtime.cc:558]   | sysTid=2332 nice=0 cgrp=default sched=0/0 handle=0xf6b54494
F webview_zygote: runtime.cc:558]   | state=R schedstat=( 102025203 15767659 98 ) utm=7 stm=3 core=6 HZ=100
F webview_zygote: runtime.cc:558]   | stack=0xff72b000-0xff72d000 stackSize=8MB
F webview_zygote: runtime.cc:558]   | held mutexes= "abort lock"
F webview_zygote: runtime.cc:558]   native: #00 pc 002d975f  /system/lib/libart.so (art::DumpNativeStack(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, int, BacktraceMap*, char const*, art::ArtMethod*, void*, bool)+134)
F webview_zygote: runtime.cc:558]   native: #01 pc 0036e98b  /system/lib/libart.so (art::Thread::DumpStack(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, bool,BacktraceMap*, bool) const+210)
F webview_zygote: runtime.cc:558]   native: #02 pc 0036b143  /system/lib/libart.so (art::Thread::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, bool, BacktraceMap*, bool) const+34)
F webview_zygote: runtime.cc:558]   native: #03 pc 003598b3  /system/lib/libart.so (art::AbortState::DumpThread(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, art::Thread*) const+30)
F webview_zygote: runtime.cc:558]   native: #04 pc 0034d953  /system/lib/libart.so (art::Runtime::Abort(char const*)+402)
F webview_zygote: runtime.cc:558]   native: #05 pc 000071b3  /system/lib/libbase.so (android::base::LogMessage::~LogMessage()+494)
F webview_zygote: runtime.cc:558]   native: #06 pc 00265a0f  /system/lib/libart.so (art::JNI::FatalError(_JNIEnv*, char const*)+122)
F webview_zygote: runtime.cc:558]   native: #07 pc 0010b4d5  /system/lib/libandroid_runtime.so (_ZZN12_GLOBAL__N_123ForkAndSpecializeCommonEP7_JNIEnvjjP10_jintArrayiP13_jobjectArrayxxiP8_jstringS7_bS3_S3_bS7_S7_ENK3$_0clERKNSt3__112basic_stringIcNS9_11char_traitsIcEENS9_9allocatorIcEEEE+92)
F webview_zygote: runtime.cc:558]   native: #08 pc 0010b28f  /system/lib/libandroid_runtime.so ((anonymous namespace)::ForkAndSpecializeCommon(_JNIEnv*, unsigned int, unsigned int, _jintArray*, int, _jobjectArray*, long long, long long, int, _jstring*, _jstring*, bool, _jintArray*, _jintArray*, bool, _jstring*, _jstring*)+4958)
F webview_zygote: runtime.cc:558]   native: #09 pc 001097f1  /system/lib/libandroid_runtime.so (android::com_android_internal_os_Zygote_nativeForkAndSpecialize(_JNIEnv*, _jclass*, int, int, _jintArray*, int, _jobjectArray*, int, _jstring*, _jstring*, _jintArray*, _jintArray*, unsigned char, _jstring*, _jstring*)+476)
F webview_zygote: runtime.cc:558]   native: #10 pc 003cb633  /system/framework/arm/boot-framework.oat (offset 3ab000) (com.android.internal.os.Zygote.nativeForkAndSpecialize+338)
F webview_zygote: runtime.cc:558]   native: #11 pc 00a0ba7f  /system/framework/arm/boot-framework.oat (offset 3ab000) (com.android.internal.os.Zygote.forkAndSpecialize+198)
F webview_zygote: runtime.cc:558]   native: #12 pc 00a0eb0d  /system/framework/arm/boot-framework.oat (offset 3ab000) (com.android.internal.os.ZygoteConnection.processOneCommand+1388)
F webview_zygote: runtime.cc:558]   native: #13 pc 00a13675  /system/framework/arm/boot-framework.oat (offset 3ab000) (com.android.internal.os.ZygoteServer.runSelectLoop+780)
F webview_zygote: runtime.cc:558]   native: #14 pc 0040d575  /system/lib/libart.so (art_quick_invoke_stub_internal+68)
F webview_zygote: runtime.cc:558]   native: #15 pc 003e6b79  /system/lib/libart.so (art_quick_invoke_stub+224)
F webview_zygote: runtime.cc:558]   native: #16 pc 000a1015  /system/lib/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+136)
F webview_zygote: runtime.cc:558]   native: #17 pc 001e5ae9  /system/lib/libart.so (art::interpreter::ArtInterpreterToCompiledCodeBridge(art::Thread*, art::ArtMethod*, art::ShadowFrame*, unsigned short, art::JValue*)+236)
F webview_zygote: runtime.cc:558]   native: #18 pc 001e05d7  /system/lib/libart.so (bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*)+814)
F webview_zygote: runtime.cc:558]   native: #19 pc 003e3c1d  /system/lib/libart.so (MterpInvokeVirtualQuick+428)
F webview_zygote: runtime.cc:558]   native: #20 pc 00404094  /system/lib/libart.so (ExecuteMterpImpl+29972)
F webview_zygote: runtime.cc:558]   native: #21 pc 00dc4ff8  /system/framework/boot-framework.vdex (com.android.internal.os.WebViewZygoteInit.main+198)
F webview_zygote: runtime.cc:558]   native: #22 pc 001c4d53  /system/lib/libart.so (_ZN3art11interpreterL7ExecuteEPNS_6ThreadERKNS_20CodeItemDataAccessorERNS_11ShadowFrameENS_6JValueEb.llvm.2471763592+378)
F webview_zygote: runtime.cc:558]   native: #23 pc 001c937f  /system/lib/libart.so (art::interpreter::EnterInterpreterFromEntryPoint(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*)+82)
F webview_zygote: runtime.cc:558]   native: #24 pc 003d52b9  /system/lib/libart.so (artQuickToInterpreterBridge+880)
F webview_zygote: runtime.cc:558]   native: #25 pc 00411aff  /system/lib/libart.so (art_quick_to_interpreter_bridge+30)
I WifiService: acquireMulticastLock uid=10087
F webview_zygote: runtime.cc:558]   native: #26 pc 0040d575  /system/lib/libart.so (art_quick_invoke_stub_internal+68)
F webview_zygote: runtime.cc:558]   native: #27 pc 003e6c7b  /system/lib/libart.so (art_quick_invoke_static_stub+222)
F webview_zygote: runtime.cc:558]   native: #28 pc 000a1027  /system/lib/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+154)
F webview_zygote: runtime.cc:558]   native: #29 pc 00347ac5  /system/lib/libart.so (art::(anonymous namespace)::InvokeWithArgArray(art::ScopedObjectAccessAlreadyRunnable const&, art::ArtMethod*, art::(anonymous namespace)::ArgArray*, art::JValue*, char const*)+52)
F webview_zygote: runtime.cc:558]   native: #30 pc 00348f15  /system/lib/libart.so (art::InvokeMethod(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jobject*, _jobject*, unsigned int)+1024)
F webview_zygote: runtime.cc:558]   native: #31 pc 002fb0c5  /system/lib/libart.so (art::Method_invoke(_JNIEnv*, _jobject*, _jobject*, _jobjectArray*)+40)
F webview_zygote: runtime.cc:558]   native: #32 pc 0011226f  /system/framework/arm/boot.oat (offset 10c000) (java.lang.Class.getDeclaredMethodInternal [DEDUPED]+110)
W ctxmgr  : [PendingIntentCompat]Timed out delivering to pendingIntent=PendingIntent{8dc1703: android.os.BinderProxy@4c24eee}, intent=Intent { (has extras) }, permission=null
F webview_zygote: runtime.cc:558]   native: #33 pc 00a0a95b  /system/framework/arm/boot-framework.oat (offset 3ab000) (com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run+114)
F webview_zygote: runtime.cc:558]   native: #34 pc 00a10845  /system/framework/arm/boot-framework.oat (offset 3ab000) (com.android.internal.os.ZygoteInit.main+2836)
F webview_zygote: runtime.cc:558]   native: #35 pc 0040d575  /system/lib/libart.so (art_quick_invoke_stub_internal+68)
F webview_zygote: runtime.cc:558]   native: #36 pc 003e6c7b  /system/lib/libart.so (art_quick_invoke_static_stub+222)
F webview_zygote: runtime.cc:558]   native: #37 pc 000a1027  /system/lib/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+154)
F webview_zygote: runtime.cc:558]   native: #38 pc 00347ac5  /system/lib/libart.so (art::(anonymous namespace)::InvokeWithArgArray(art::ScopedObjectAccessAlreadyRunnable const&, art::ArtMethod*, art::(anonymous namespace)::ArgArray*, art::JValue*, char const*)+52)
F webview_zygote: runtime.cc:558]   native: #39 pc 003478ef  /system/lib/libart.so (art::InvokeWithVarArgs(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jmethodID*, std::__va_list)+310)
F webview_zygote: runtime.cc:558]   native: #40 pc 0028eb11  /system/lib/libart.so (art::JNI::CallStaticVoidMethodV(_JNIEnv*, _jclass*, _jmethodID*, std::__va_list)+444)
F webview_zygote: runtime.cc:558]   native: #41 pc 0006c99b  /system/lib/libandroid_runtime.so (_JNIEnv::CallStaticVoidMethod(_jclass*, _jmethodID*, ...)+30)
F webview_zygote: runtime.cc:558]   native: #42 pc 0006ebf3  /system/lib/libandroid_runtime.so (android::AndroidRuntime::start(char const*, android::Vector<android::String8>

F webview_zygote: runtime.cc:558]   native: #43 pc 00001989  /system/bin/app_process32 (???)
F webview_zygote: runtime.cc:558]   native: #44 pc 0008b89d  /system/lib/libc.so (__libc_init+48)
F webview_zygote: runtime.cc:558]   native: #45 pc 0000166f  /system/bin/app_process32 (???)
F webview_zygote: runtime.cc:558]   native: #46 pc 00000306  <anonymous:f6b51000> (???)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.Zygote.nativeForkAndSpecialize(Native method)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.Zygote.forkAndSpecialize(Zygote.java:139)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.ZygoteConnection.processOneCommand(ZygoteConnection.java:234)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.ZygoteServer.runSelectLoop(ZygoteServer.java:204)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.WebViewZygoteInit.main(WebViewZygoteInit.java:160)
F webview_zygote: runtime.cc:558]   at java.lang.reflect.Method.invoke(Native method)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
F webview_zygote: runtime.cc:558]   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
F webview_zygote: runtime.cc:558]
E libc    : failed to connect to tombstoned: Permission denied
W webview_zygote: type=1400 audit(0.0:30): avc: denied { write } for name="tombstoned_crash" dev="tmpfs" ino=16442 scontext=u:r:webview_zygote:s0 tcontext=u:object_r:tombstoned_crash_socket:s0 tclass=sock_file permissive=0

F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
F DEBUG   : ABI: 'arm'
F DEBUG   : pid: 2332, tid: 2332, name: webview_zygote  >>> webview_zygote <<<
F DEBUG   : signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
F DEBUG   : Abort message: 'jni_internal.cc:616] JNI FatalError called: (com.android.webview:sandboxed_process0) Unable to stat 28'
F DEBUG   :     r0  00000000  r1  0000091c  r2  00000006  r3  00000008
F DEBUG   :     r4  0000091c  r5  0000091c  r6  fff24e2c  r7  0000010c
F DEBUG   :     r8  0000000d  r9  f2f1c720  r10 fff24f80  r11 f2e2763e
F DEBUG   :     ip  fff24dc8  sp  fff24e18  lr  f502cdd9  pc  f5024c7a
D Fabric  : Using AdvertisingInfo from Preference Store
I Gecko   : 1538430036192        addons.xpi      WARN    List of valid built-in add-ons could not be parsed.: [Exception... "Component returned failure code: 0x80520001 (NS_ERROR_FILE_UNRECOGNIZED_PATH) [nsIXPCComponents_Utils.readUTF8URI]"  nsresult: "0x80520001 (NS_ERROR_FILE_UNRECOGNIZED_PATH)"  location: "JS frame :: resource://gre/modules/addons/XPIProvider.jsm :: startup :: line 197"  data: no] Stack trace: startup()@resource://gre/modules/addons/XPIProvider.jsm:197
W webview_zygote: type=1400 audit(0.0:31): avc: denied { read } for name="app_process32" dev="mmcblk0p25" ino=467 scontext=u:r:webview_zygote:s0 tcontext=u:object_r:zygote_exec:s0 tclass=file permissive=0
I unwind  : Malformed section header found, ignoring...
F DEBUG   :
F DEBUG   : backtrace:
F DEBUG   :     #00 pc 0001cc7a  /system/lib/libc.so (abort+58)
F DEBUG   :     #01 pc 0034db4f  /system/lib/libart.so (art::Runtime::Abort(char const*)+910)
F DEBUG   :     #02 pc 000071b3  /system/lib/libbase.so (android::base::LogMessage::~LogMessage()+494)
F DEBUG   :     #03 pc 00265a0f  /system/lib/libart.so (art::JNI::FatalError(_JNIEnv*, char const*)+122)
F DEBUG   :     #04 pc 0010b4d5  /system/lib/libandroid_runtime.so (_ZZN12_GLOBAL__N_123ForkAndSpecializeCommonEP7_JNIEnvjjP10_jintArrayiP13_jobjectArrayxxiP8_jstringS7_bS3_S3_bS7_S7_ENK3$_0clERKNSt3__112basic_stringIcNS9_11char_traitsIcEENS9_9allocatorIcEEEE+92)
F DEBUG   :     #05 pc 0010b28f  /system/lib/libandroid_runtime.so ((anonymous namespace)::ForkAndSpecializeCommon(_JNIEnv*, unsigned int, unsigned int, _jintArray*, int, _jobjectArray*, long long, long long, int, _jstring*, _jstring*, bool, _jintArray*, _jintArray*, bool, _jstring*, _jstring*)+4958)
F DEBUG   :     #06 pc 001097f1  /system/lib/libandroid_runtime.so (android::com_android_internal_os_Zygote_nativeForkAndSpecialize(_JNIEnv*, _jclass*, int, int, _jintArray*, int, _jobjectArray*, int, _jstring*, _jstring*, _jintArray*, _jintArray*, unsigned char, _jstring*, _jstring*)+476)
F DEBUG   :     #07 pc 003cb633  /system/framework/arm/boot-framework.oat (offset 0x3ab000) (com.android.internal.os.Zygote.nativeForkAndSpecialize+338)
F DEBUG   :     #08 pc 00a0ba7f  /system/framework/arm/boot-framework.oat (offset 0x3ab000) (com.android.internal.os.Zygote.forkAndSpecialize+198)
F DEBUG   :     #09 pc 00a0eb0d  /system/framework/arm/boot-framework.oat (offset 0x3ab000) (com.android.internal.os.ZygoteConnection.processOneCommand+1388)
F DEBUG   :     #10 pc 00a13675  /system/framework/arm/boot-framework.oat (offset 0x3ab000) (com.android.internal.os.ZygoteServer.runSelectLoop+780)
F DEBUG   :     #11 pc 0040d575  /system/lib/libart.so (art_quick_invoke_stub_internal+68)
F DEBUG   :     #12 pc 003e6b79  /system/lib/libart.so (art_quick_invoke_stub+224)
F DEBUG   :     #13 pc 000a1015  /system/lib/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+136)
F DEBUG   :     #14 pc 001e5ae9  /system/lib/libart.so (art::interpreter::ArtInterpreterToCompiledCodeBridge(art::Thread*, art::ArtMethod*, art::ShadowFrame*, unsigned short, art::JValue*)+236)
F DEBUG   :     #15 pc 001e05d7  /system/lib/libart.so (bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*)+814)
F DEBUG   :     #16 pc 003e3c1d  /system/lib/libart.so (MterpInvokeVirtualQuick+428)
F DEBUG   :     #17 pc 00404094  /system/lib/libart.so (ExecuteMterpImpl+29972)
F DEBUG   :     #18 pc 00dc4ff8  /system/framework/boot-framework.vdex (com.android.internal.os.WebViewZygoteInit.main+198)
F DEBUG   :     #19 pc 001c4d53  /system/lib/libart.so (_ZN3art11interpreterL7ExecuteEPNS_6ThreadERKNS_20CodeItemDataAccessorERNS_11ShadowFrameENS_6JValueEb.llvm.2471763592+378)
F DEBUG   :     #20 pc 001c937f  /system/lib/libart.so (art::interpreter::EnterInterpreterFromEntryPoint(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*)+82)
F DEBUG   :     #21 pc 003d52b9  /system/lib/libart.so (artQuickToInterpreterBridge+880)
F DEBUG   :     #22 pc 00411aff  /system/lib/libart.so (art_quick_to_interpreter_bridge+30)
F DEBUG   :     #23 pc 0040d575  /system/lib/libart.so (art_quick_invoke_stub_internal+68)
F DEBUG   :     #24 pc 003e6c7b  /system/lib/libart.so (art_quick_invoke_static_stub+222)
F DEBUG   :     #25 pc 000a1027  /system/lib/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+154)
F DEBUG   :     #26 pc 00347ac5  /system/lib/libart.so (art::(anonymous namespace)::InvokeWithArgArray(art::ScopedObjectAccessAlreadyRunnable const&, art::ArtMethod*, art::(anonymous namespace)::ArgArray*, art::JValue*, char const*)+52)
F DEBUG   :     #27 pc 00348f15  /system/lib/libart.so (art::InvokeMethod(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jobject*, _jobject*, unsigned int)+1024)
F DEBUG   :     #28 pc 002fb0c5  /system/lib/libart.so (art::Method_invoke(_JNIEnv*, _jobject*, _jobject*, _jobjectArray*)+40)
F DEBUG   :     #29 pc 0011226f  /system/framework/arm/boot.oat (offset 0x10c000) (java.lang.Class.getDeclaredMethodInternal [DEDUPED]+110)
F DEBUG   :     #30 pc 00a0a95b  /system/framework/arm/boot-framework.oat (offset 0x3ab000) (com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run+114)
F DEBUG   :     #31 pc 00a10845  /system/framework/arm/boot-framework.oat (offset 0x3ab000) (com.android.internal.os.ZygoteInit.main+2836)
F DEBUG   :     #32 pc 0040d575  /system/lib/libart.so (art_quick_invoke_stub_internal+68)
F DEBUG   :     #33 pc 003e6c7b  /system/lib/libart.so (art_quick_invoke_static_stub+222)
F DEBUG   :     #34 pc 000a1027  /system/lib/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+154)
F DEBUG   :     #35 pc 00347ac5  /system/lib/libart.so (art::(anonymous namespace)::InvokeWithArgArray(art::ScopedObjectAccessAlreadyRunnable const&, art::ArtMethod*, art::(anonymous namespace)::ArgArray*, art::JValue*, char const*)+52)
F DEBUG   :     #36 pc 003478ef  /system/lib/libart.so (art::InvokeWithVarArgs(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jmethodID*, std::__va_list)+310)
F DEBUG   :     #37 pc 0028eb11  /system/lib/libart.so (art::JNI::CallStaticVoidMethodV(_JNIEnv*, _jclass*, _jmethodID*, std::__va_list)+444)
F DEBUG   :     #38 pc 0006c99b  /system/lib/libandroid_runtime.so (_JNIEnv::CallStaticVoidMethod(_jclass*, _jmethodID*, ...)+30)
F DEBUG   :     #39 pc 0006ebf3  /system/lib/libandroid_runtime.so (android::AndroidRuntime::start(char const*, android::Vector<android::String8> const&, bool)+458)
F DEBUG   :     #40 pc 00001989  /system/bin/app_process32
F DEBUG   :     #41 pc 0008b89d  /system/lib/libc.so (__libc_init+48)
F DEBUG   :     #42 pc 0000166f  /system/bin/app_process32
F DEBUG   :     #43 pc 00000306  <anonymous:f6b51000>
W webview_zygote: type=1400 audit(0.0:32): avc: denied { read } for name="app_process32" dev="mmcblk0p25" ino=467 scontext=u:r:webview_zygote:s0 tcontext=u:object_r:zygote_exec:s0 tclass=file permissive=0
E ZygoteProcess: Starting VM process through Zygote failed
I Zygote  : Process 2332 exited due to signal (6)
E ActivityManager: Failure starting process com.android.webview:sandboxed_process0
E ActivityManager: java.lang.RuntimeException: Starting VM process through Zygote failed
E ActivityManager:       at android.os.ZygoteProcess.start(ZygoteProcess.java:239)
E ActivityManager:       at android.os.Process.startWebView(Process.java:507)
E ActivityManager:       at com.android.server.am.ActivityManagerService.startProcess(ActivityManagerService.java:4478)
E ActivityManager:       at com.android.server.am.ActivityManagerService.lambda$startProcessLocked$0(ActivityManagerService.java:4432)
E ActivityManager:       at com.android.server.am.-$$Lambda$ActivityManagerService$UgpguyCBuObHjnmry_xkrJGkFi0.run(Unknown Source:20)
E ActivityManager:       at android.os.Handler.handleCallback(Handler.java:873)
E ActivityManager:       at android.os.Handler.dispatchMessage(Handler.java:99)
E ActivityManager:       at android.os.Looper.loop(Looper.java:193)
E ActivityManager:       at android.os.HandlerThread.run(HandlerThread.java:65)
E ActivityManager:       at com.android.server.ServiceThread.run(ServiceThread.java:44)
E ActivityManager: Caused by: android.os.ZygoteStartFailedEx: java.io.EOFException
E ActivityManager:       at android.os.ZygoteProcess.zygoteSendArgsAndGetResult(ZygoteProcess.java:332)
E ActivityManager:       at android.os.ZygoteProcess.startViaZygote(ZygoteProcess.java:438)
E ActivityManager:       at android.os.ZygoteProcess.start(ZygoteProcess.java:232)
E ActivityManager:       ... 9 more
E ActivityManager: Caused by: java.io.EOFException
E ActivityManager:       at java.io.DataInputStream.readFully(DataInputStream.java:200)
E ActivityManager:       at java.io.DataInputStream.readInt(DataInputStream.java:389)
E ActivityManager:       at android.os.ZygoteProcess.zygoteSendArgsAndGetResult(ZygoteProcess.java:323)
E ActivityManager:       ... 11 more
I ActivityManager: Force stopping com.google.android.gm appid=99000 user=0: start failure
I ActivityManager:   Force finishing activity ActivityRecord{ad8615d u0 com.google.android.gm/.ConversationListActivityGmail t116}
E ActivityManager: Found activity ActivityRecord{ad8615d u0 com.google.android.gm/.ConversationListActivityGmail t116 f} in proc activity list using null instead of expected ProcessRecord{f4179d8 4833:com.google.android.gm/u0a49}

Change-Id: I7eadf4ac2f286c90a92cd84b26b00f0f29b7a57e
Signed-off-by: Vladimir Oltean <olteanv@gmail.com>
---
 private/webview_zygote.te | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/private/webview_zygote.te b/private/webview_zygote.te
index 55b268a3..9b5f7060 100644
--- a/private/webview_zygote.te
+++ b/private/webview_zygote.te
@@ -51,6 +51,8 @@ dontaudit webview_zygote mnt_expand_file:dir getattr;
 # app_process zygote
 dontaudit webview_zygote dex2oat_exec:file execute;
 
+allow webview_zygote ion_device:chr_file r_file_perms;
+
 # Get seapp_contexts
 allow webview_zygote seapp_contexts_file:file r_file_perms;
 # Check validity of SELinux context before use.
-- 
2.17.1

